packages/libraries/router/src/persisted_documents.rs

```diff
@@ -0,0 +1,790 @@
+    #[error("Failed to parse body: {0}")]
+    FailedToParseBody(serde_json::Error),
+    #[error("Persisted document not found.")]
+    DocumentNotFound,
+    #[error("Failed to locate the persisted document key in request.")]
+    KeyNotFound,
+    #[error("Failed to validate persisted document")]
+    FailedToFetchFromCDN(reqwest::Error),
+    #[error("No persisted document provided, or document id cannot be resolved.")]
+    PersistedDocumentRequired,
+}
+
+impl PersistedDocumentsError {
+    fn message(&self) -> String {
+        self.to_string()
+    }
+
+    fn code(&self) -> String {
+        match self {
+            PersistedDocumentsError::FailedToReadBody(_) => "FAILED_TO_READ_BODY".into(),
+            PersistedDocumentsError::FailedToParseBody(_) => "FAILED_TO_PARSE_BODY".into(),
+            PersistedDocumentsError::DocumentNotFound => "PERSISTED_DOCUMENT_NOT_FOUND".into(),
+            PersistedDocumentsError::KeyNotFound => "PERSISTED_DOCUMENT_KEY_NOT_FOUND".into(),
+            PersistedDocumentsError::FailedToFetchFromCDN(_) => "FAILED_TO_FETCH_FROM_CDN".into(),
+            PersistedDocumentsError::PersistedDocumentRequired => {
+                "PERSISTED_DOCUMENT_REQUIRED".into()
+            }
+        }
+    }
+
+    fn to_router_response(&self, ctx: Context) -> router::Response {
+        let errors = vec![Error::builder()
+            .message(self.message())
+            .extension_code(self.code())
+            .build()];
+
+        router::Response::error_builder()
+            .errors(errors)
+            .status_code(StatusCode::OK)
+            .context(ctx)
+            .build()
+            .unwrap()
+    }
+}
+
+impl PersistedDocumentsManager {
+    fn new(config: &Config) -> Self {
+        let agent = reqwest::Client::builder()
+            .danger_accept_invalid_certs(config.accept_invalid_certs.unwrap_or(false))
+            .connect_timeout(Duration::from_secs(config.connect_timeout.unwrap_or(5)))
+            .timeout(Duration::from_secs(config.request_timeout.unwrap_or(15)))
+            .build()
+            .expect("Failed to create reqwest client");
+
+        let cache_size = config.cache_size.unwrap_or(1000);
+        let cache = Arc::new(Mutex::new(LruCache::<String, String>::new(
+            NonZeroUsize::new(cache_size).unwrap(),
+        )));
+
+        Self {
+            agent,
+            cache,
+            config: config.clone(),
+        }
+    }
+
+    /// Extracts the document id from the request body.
+    /// In case of a parsing error, it returns the error.
+    /// This will also try to parse other GraphQL-related (see `original_req`) fields in order to 
+    /// pass it to the next layer.
+    fn extract_document_id(
+        body: &hyper::body::Bytes,
+    ) -> Result<ExpectedBodyStructure, PersistedDocumentsError> {
+        serde_json::from_slice::<ExpectedBodyStructure>(&body)
+            .map_err(PersistedDocumentsError::FailedToParseBody)
```

Comment:
there's no differentiation between a malformed JSON body and a missing document ID, maybe there should be more variants for clearer error messages:
- `DOCUMENT_ID_MISSING`
- `MALFORMED_REQUEST_BODY`

Reply:
Actually, the `structure ExpectedBodyStructure` is defined with `document_id: Option<String>` field, so it will not fail in case of a missing document id. 
The document ID is checked later for its value and then a specific error is returned. 
So this one can actually fail only in a case the json parsing fails.